src/mscorlib/src/System/Globalization/CompareInfo.Invariant.cs

```diff
@@ -0,0 +1,238 @@
+        {
+            Debug.Assert(source != null);
+            Debug.Assert(value != null);
+            Debug.Assert(startIndex >= 0 && startIndex < source.Length);
+
+            fixed (char* pSource = source) fixed (char* pValue = value)
+            {
+                char* pSrc = &pSource[startIndex - count + 1];
+                int index = InvariantFindString(pSrc, count, pValue, value.Length, ignoreCase, start : false);
+                if (index >= 0)
+                {
+                    return index + startIndex - count + 1;
+                }
+                return -1;
+            }
+        }
+
+        private static unsafe int InvariantFindString(char* source, int sourceCount, char* value, int valueCount, bool ignoreCase, bool start)
+        {
+            int ctrSource = 0;  // index value into source
+            int ctrValue = 0;   // index value into value
+            char sourceChar;    // Character for case lookup in source
+            char valueChar;     // Character for case lookup in value
+            int lastSourceStart;
+
+            Debug.Assert(source != null);
+            Debug.Assert(value != null);
+            Debug.Assert(sourceCount >= 0);
+            Debug.Assert(valueCount >= 0);
+
+            if (valueCount == 0)
+            {
+                return start ? 0 : sourceCount - 1;
+            }
+
+            if (sourceCount < valueCount)
+            {
+                return -1;
+            }
+
+            if (start)
+            {
+                lastSourceStart = sourceCount - valueCount;
+                if (ignoreCase)
+                {
+                    char firstValueChar = InvariantToUpper(value[0]);
+                    for (ctrSource = 0; ctrSource <= lastSourceStart; ctrSource++)
+                    {
+                        sourceChar = InvariantToUpper(source[ctrSource]);
+                        if (sourceChar != firstValueChar)
+                        {
+                            continue;
+                        }
+
+                        for (ctrValue = 1; ctrValue < valueCount; ctrValue++)
+                        {
+                            sourceChar = InvariantToUpper(source[ctrSource + ctrValue]);
+                            valueChar = InvariantToUpper(value[ctrValue]);
+
+                            if (sourceChar != valueChar)
+                            {
+                                break;
+                            }
+                        }
+
+                        if (ctrValue == valueCount)
+                        {
+                            return ctrSource;
+                        }
+                    }
+                }
+                else
+                {
+                    char firstValueChar = value[0];
+                    for (ctrSource = 0; ctrSource <= lastSourceStart; ctrSource++)
+                    {
+                        sourceChar = source[ctrSource];
```

Comment:
Why special case the first character instead of just 
```c#
                else
                {
                    for (ctrSource = 0; ctrSource <= lastSourceStart; ctrSource++)
                    {
                        for (ctrValue = 0; ctrValue < valueCount; ctrValue++)
                        {
                            sourceChar = source[ctrSource + ctrValue];
                            valueChar = value[ctrValue];

                            if (sourceChar != valueChar)
                            {
                                break;
                            }
                        }
...
```

In the two case insensitive blocks, you've special cased the first character of the source to avoid having to ToUpperCase it more than once, but in the case sensitive blocks, there's no value is there?